Observer/TrackingSaveObserver.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace Bolt\Boltpay\Observer;
+
+use Bolt\Boltpay\Helper\Api as ApiHelper;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\Framework\Event\Observer;
+use Bolt\Boltpay\Helper\Config as ConfigHelper;
+use Magento\Framework\DataObjectFactory;
+
+/**
+ * Class TrackingSaveObserver
+ *
+ * @package Bolt\Boltpay\Observer
+ */
+class TrackingSaveObserver implements ObserverInterface
+{
+    /**
+     * @var ConfigHelper
+     */
+    private $configHelper;
+
+    /**
+     * @var DataObjectFactory
+     */
+    private $dataObjectFactory;
+
+    /**
+     * @param ConfigHelper      $configHelper
+     * @param DataObjectFactory $dataObjectFactory
+     *
+     */
+    public function __construct(
+        ConfigHelper $configHelper,
+        DataObjectFactory $dataObjectFactory
+    ) {
+        $this->configHelper = $configHelper;
+        $this->dataObjectFactory = $dataObjectFactory;
+    }
+
+    /**
+     * @param Observer $observer
+     */
+    public function execute(Observer $observer)
+    {
+        $tracking = $observer->getEvent()->getTrack();
+        $shipment = $tracking->getShipment();
+        $order = $shipment->getOrder();
+        $payment = $order->getPayment();
+        $transactionReference = $payment->getAdditionalInformation('transaction_reference');
+        $items = [];
+        foreach ($shipment->getItemsCollection() as $item) {
+            $items[] = $item->getOrderItem()->getProductId();
+        }
+
+        $apiKey = $this->configHelper->getApiKey($order->getStoreId());
+
+        $trackingData = [
+            'transaction_reference' => $transactionReference,
+            'tracking_number'       => $tracking->getTrackNumber(),
+            'carrier'               => $tracking->getCarrierCode(),
+            'items'                 => $items
+        ];
+
+        //Request Data
+        $requestData = $this->dataObjectFactory->create();
+        $requestData->setApiData($trackingData);
+        $requestData->setDynamicApiUrl(ApiHelper::API_CREATE_TRACKING);
+        $requestData->setApiKey($apiKey);
+
+        //Build Request
+        $request = $this->apiHelper->buildRequest($requestData);
+        $this->apiHelper->sendRequest($request);
```

Comment:
I see a potential problem here.
Observer execution in Magento is sequential (synchronous) and so is this call (via Zend_Http_Client). Side effects:
1. waiting for response from Bolt in order to continue Magento processing
2. breaks the Magento flow in case of an error (exception thrown in \Bolt\Boltpay\Helper\Api::sendRequest)
Depending on how important this data is we could:
1. find a way to send asynchronous request or
2. wrap it in try / catch and decide how to approach the error
3. maybe we can do this similar to how the metrics is sent - log requests to file and send batches on interval, leaving  unsent on error

Reply:
I'm new to the codebase and so I could totally be wrong, but why would this cause an issue if there's an error? We're just subscribing to an event so I would assume there would be no side effects even if an error ocurred

Reply:
@akulkapoor I was trying to find some documentation regarding event dispatching and observer execution in M2 but could not find any real quick to support my point. So, based on experience, events can be fired somewhere, anywhere in the code and at that point a sequence of observers start to execute, synchronously, one by one, until the last one finishes and the execution continues from where it was "interrupted". If an exception is thrown and not caught in an observer (or around the the event) the execution stops.
@matt-thomason @YevhenShBolt @johnmccombs1 @njfaries please correct me if I am wrong.

Reply:
@akulkapoor I agree with Daniel. If there is an error thrown, the execution will break. I think we should at least add a try/catch to keep from breaking the execution in case there is an exception thrown in the new event that you added.

Reply:
got it, will add something

Reply:
thanks for the explanation!